### ISSUE TITLE
No match of right hand side value in typechecker:solve_bounds/4

### ISSUE THREAD

**Issue by erszcz** [josefs/Gradualizer#106]
Checking [the following file](https://github.com/erszcz/docsh/blob/db6fe56e1bcd880a0b33d0facd45447b13b29625/src/docsh_internal.erl):

```erlang
-module(docsh_internal).

-export([merge/1,
         kna/1]).

-export_type([t/0,
              item/0,
              kna/0]).

-type t()    :: #{name        := module(),
                  items       := [item()],
                  description => none | iodata()}.

-type item() :: #{kind        := function | type,
                  name        := atom(),
                  arity       := arity(),
                  description => none | iodata(),
                  exported    => boolean(),
                  signature   => iodata()}.

-type kna() :: {function | type, atom(), arity()}.

-define(a2b(A), atom_to_binary(A, utf8)).
-define(a2l(A), atom_to_list(A)).
-define(i2b(I), integer_to_binary(I)).
-define(il2b(IOList), iolist_to_binary(IOList)).

%%
%%' Public
%%

-spec merge([Info]) -> MergedInfo when
      Info :: t(),
      MergedInfo :: t().
merge([]) -> [];
merge([Info]) -> Info;
merge([Info1, Info2 | Rest]) ->
    case are_disjoint(Info1, Info2) of
        false -> erlang:error(not_disjoint, [Info1, Info2 | Rest]);
        true ->
            (module(Info1) =:= module(Info2)
             orelse erlang:error(different_modules, [Info1, Info2 | Rest])),
            merge([merge_two(Info1, Info2) | Rest])
    end.

-spec kna(item()) -> kna().
kna(#{kind := K, name := N, arity := A}) -> {K, N, A}.

%%.
%%' Internal
%%

-spec merge_two(t(), t()) -> t().
merge_two(#{items := Items1} = Info1, #{items := Items2}) ->
    ItemsByKNA = dict:to_list( docsh_lib:group_by(fun kna/1, Items1 ++ Items2) ),
    NewItems = lists:map(fun merge_two_/1,  ItemsByKNA),
    Info1#{items := NewItems}.

-spec merge_two_({kna(), [item()]}) -> item().
merge_two_({{Kind, Name, Arity}, [Item]}) ->
    #{kind := Kind, name := Name, arity := Arity} = Item,
    Item;
merge_two_({{Kind, Name, Arity}, [Item1, Item2]}) ->
    #{kind := Kind, name := Name, arity := Arity} = Item1,
    #{kind := Kind, name := Name, arity := Arity} = Item2,
    maps:merge(Item1, Item2).

are_disjoint(Info1, Info2) ->
    Module = maps:get(name, Info1),
    Module = maps:get(name, Info2),
    #{items := Items1} = Info1,
    #{items := Items2} = Info2,
    Items1 -- Items2 == Items1.

module(#{name := Name}) -> Name.

%%. vim: foldmethod=marker foldmarker=%%',%%.
```

I've run into the following error:

```
$ gradualizer src/docsh_internal.erl
src/docsh_internal.erl: The empty list on line 35 does not have type #{name := module(),
            items := [item()],
            description => none | iodata()}
escript: exception error: no match of right hand side value
                 #{'Dict' =>
                       [{user_type,[{file,"dict.erl"},{location,100}],
                                   dict,
                                   [{var,100,'Key'},{var,100,'Value'}]}],
                   'List' =>
                       [{type,101,list,
                              [{type,101,tuple,
                                     [{var,101,'Key'},{var,101,'Value'}]}]}]}
  in function  typechecker:solve_bounds/4 (/Users/erszcz/work/josefs/gradualizer/_build/default/lib/gradualizer/src/typechecker.erl, line 1182)
  in call from typechecker:bounded_type_subst/2 (/Users/erszcz/work/josefs/gradualizer/_build/default/lib/gradualizer/src/typechecker.erl, line 1166)
  in call from typechecker:expect_fun_type1/2 (/Users/erszcz/work/josefs/gradualizer/_build/default/lib/gradualizer/src/typechecker.erl, line 1037)
  in call from typechecker:expect_intersection_type/2 (/Users/erszcz/work/josefs/gradualizer/_build/default/lib/gradualizer/src/typechecker.erl, line 1095)
  in call from typechecker:expect_fun_type1/2 (/Users/erszcz/work/josefs/gradualizer/_build/default/lib/gradualizer/src/typechecker.erl, line 1058)
  in call from typechecker:expect_fun_type/2 (/Users/erszcz/work/josefs/gradualizer/_build/default/lib/gradualizer/src/typechecker.erl, line 1024)
  in call from typechecker:type_check_expr/2 (/Users/erszcz/work/josefs/gradualizer/_build/default/lib/gradualizer/src/typechecker.erl, line 1315)
  in call from typechecker:type_check_expr/2 (/Users/erszcz/work/josefs/gradualizer/_build/default/lib/gradualizer/src/typechecker.erl, line 1234)
```

A trace on `solve_bounds` gives:

```erlang
8> (<0.80.0>) call typechecker:solve_bounds({tenv,#{{item,0} =>
            {[],
             {type,14,map,
                   [{type,14,map_field_exact,
                          [{atom,14,kind},
                           {type,14,union,
                                 [{atom,14,function},{atom,14,type}]}]},
                    {type,15,map_field_exact,
                          [{atom,15,name},{type,15,atom,[]}]},
                    {type,16,map_field_exact,
                          [{atom,16,arity},{type,16,arity,[]}]},
                    {type,17,map_field_assoc,
                          [{atom,17,description},
                           {type,17,union,
                                 [{atom,17,none},{type,17,iodata,[]}]}]},
                    {type,18,map_field_assoc,
                          [{atom,18,exported},{type,18,boolean,[]}]},
                    {type,19,map_field_assoc,
                          [{atom,19,signature},{type,19,iodata,[]}]}]}},
        {kna,0} =>
            {[],
             {type,21,tuple,
                   [{type,21,union,[{atom,21,function},{atom,21,type}]},
                    {type,21,atom,[]},
                    {type,21,arity,[]}]}},
        {t,0} =>
            {[],
             {type,10,map,
                   [{type,10,map_field_exact,
                          [{atom,10,name},{type,10,module,[]}]},
                    {type,11,map_field_exact,
                          [{atom,11,items},
                           {type,11,list,[{user_type,11,item,[]}]}]},
                    {type,12,map_field_assoc,
                          [{atom,12,description},
                           {type,12,union,
                                 [{atom,12,none},{type,12,iodata,[]}]}]}]}}},
      #{}},#{'Dict' =>
      [{user_type,[{file,"dict.erl"},{location,100}],
                  dict,
                  [{var,100,'Key'},{var,100,'Value'}]}],
  'List' =>
      [{type,101,list,
             [{type,101,tuple,[{var,101,'Key'},{var,101,'Value'}]}]}]},[{acyclic,'Value'},{acyclic,'Key'},{acyclic,'List'},{acyclic,'Dict'}],#{})
(<0.80.0>) exception_from {typechecker,solve_bounds,4} {error,{badmatch,#{'Dict' =>
                       [{user_type,[{file,"dict.erl"},{location,100}],
                                   dict,
                                   [{var,100,'Key'},{var,100,'Value'}]}],
                   'List' =>
                       [{type,101,list,
                              [{type,101,tuple,
                                     [{var,101,'Key'},
                                      {var,101,'Value'}]}]}]}}}
```

My not so educated guess is that there's no complete support for maps or a bug there. Might it be the case?

**Comment by zuiderkwast** [comment]
I got this error when running gradualizer on itself and I think I have fixed it in 274e8c90f174bcd0b7a33641e71f25f71f0630f4. It looks like the same error. @erszcz can you verify?

**Comment by erszcz** [comment]
@zuiderkwast Indeed, the error is gone. Awesome!
